src/subdomains/user/api/controllers/dfx.controller.ts

```diff
@@ -0,0 +1,108 @@
+import { Body, Controller, ForbiddenException, Get, NotFoundException, Post, Query, UseGuards } from '@nestjs/common';
+import { ApiExcludeEndpoint } from '@nestjs/swagger';
+import { AuthGuard } from '@nestjs/passport';
+import { WalletService } from '../../application/services/wallet.service';
+import { Config } from 'src/config/config';
+import { RealIP } from 'nestjs-real-ip';
+import { WebhookDto, WebhookType } from '../../application/dto/webhook-data.dto';
+import { UserService } from '../../application/services/user.service';
+import { KycStatus } from '../../domain/enums';
+import { KycCompleted } from '../../domain/utils';
+import { NotificationService } from 'src/integration/notification/services/notification.service';
+import { MailType } from 'src/integration/notification/enums';
+import { KycWebhookData } from '../../application/dto/kyc-webhook.dto';
+
+@Controller('dfx')
+export class DfxController {
+  constructor(
+    private readonly walletService: WalletService,
+    private readonly userService: UserService,
+    private readonly notificationService: NotificationService,
+  ) {}
+
+  // --- DFX SERVICE--- //
+
+  @Get('check')
+  @UseGuards(AuthGuard('api-key'))
+  @ApiExcludeEndpoint()
+  async getKycId(@RealIP() ip: string, @Query('address') address: string): Promise<{ kycId: string }> {
+    this.checkIp(ip, address);
+    return { kycId: await this.walletService.getKycIdByAddress(address) };
+  }
+
+  @Post('update')
+  @UseGuards(AuthGuard('api-key'))
+  @ApiExcludeEndpoint()
+  async handoverDfxWebhook(@RealIP() ip: string, @Body() dto: WebhookDto) {
+    this.checkIp(ip, dto.data);
+
+    const user = await this.userService.getUserByKycId(dto.id);
+    if (!user) throw new NotFoundException('User not found');
+
+    switch (dto.type) {
```

Comment:
Maybe add two methods `onKycChanged` and `onKycFailed` to improve readability of the switch?